File: frontend/src/services/projectService.ts
import { apiClient } from '../api/axios';
import type { PageResponse, Project, ProjectStatus } from '../types';

export const projectService = {
  list: async (workspaceId: string, search?: string, page: number = 0, size: number = 10, signal?: AbortSignal) => {
    const params: any = { page, size };
    if (search) params.search = search;
    const response = await apiClient.get<PageResponse<Project>>(`/workspaces/${workspaceId}/projects`, { params, signal });
    return response.data;
  },

  get: async (workspaceId: string, projectId: string, signal?: AbortSignal) => {
    const response = await apiClient.get<Project>(`/workspaces/${workspaceId}/projects/${projectId}`, { signal });
    return response.data;
  },

  create: async (workspaceId: string, payload: { name: string; description: string }) => {
    const response = await apiClient.post<Project>(`/workspaces/${workspaceId}/projects`, payload);
    return response.data;
  },

  update: async (workspaceId: string, projectId: string, payload: { name: string; description: string }) => {
    const response = await apiClient.put<Project>(`/workspaces/${workspaceId}/projects/${projectId}`, payload);
    return response.data;
  },

  updateStatus: async (workspaceId: string, projectId: string, status: ProjectStatus) => {
    const response = await apiClient.patch<Project>(`/workspaces/${workspaceId}/projects/${projectId}/status`, { status });
    return response.data;
  },

  remove: async (workspaceId: string, projectId: string) => {
    await apiClient.delete(`/workspaces/${workspaceId}/projects/${projectId}`);
  }
};

File: backend/src/main/java/com/agiletrack/backend/project/service/ProjectService.java
package com.agiletrack.backend.project.service;

import com.agiletrack.backend.common.exception.BusinessRuleException;
import com.agiletrack.backend.common.exception.ProjectNotFoundException;
import com.agiletrack.backend.project.dto.CreateProjectRequest;
import com.agiletrack.backend.project.dto.ProjectResponse;
import com.agiletrack.backend.project.dto.UpdateProjectRequest;
import com.agiletrack.backend.project.dto.UpdateProjectStatusRequest;
import com.agiletrack.backend.project.entity.Project;
import com.agiletrack.backend.project.entity.ProjectStatus;
import com.agiletrack.backend.project.mapper.ProjectMapper;
import com.agiletrack.backend.project.repository.ProjectRepository;
import com.agiletrack.backend.workspace.entity.Workspace;
import com.agiletrack.backend.workspace.service.WorkspaceService;
import org.springframework.transaction.annotation.Transactional;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class ProjectService {
    private final WorkspaceService workspaceService;
    private final ProjectRepository projectRepository;
    private final ProjectMapper projectMapper;

    @Transactional
    public ProjectResponse createProject(UUID workspaceId, @Valid CreateProjectRequest request) {
        Workspace workspace = workspaceService.getWorkspaceForMutation(workspaceId);

        Project project = Project.builder()
                .name(request.name())
                .description(request.description())
                .status(ProjectStatus.PLANNING)
                .workspace(workspace)
                .build();

        project = projectRepository.save(project);
        return projectMapper.toResponse(project);
    }

    @Transactional(readOnly = true)
    public Page<ProjectResponse> getProjectsByWorkspace(UUID workspaceId, String search, Pageable pageable) {
        workspaceService.getWorkspaceIfMember(workspaceId);

        Page<Project> projects;
        if (search != null && !search.trim().isEmpty()) {
            projects = projectRepository.findByWorkspaceIdAndSearch(workspaceId, search.trim(), pageable);
        } else {
            projects = projectRepository.findByWorkspaceId(workspaceId, pageable);
        }

        return projects.map(projectMapper::toResponse);
    }

    @Transactional(readOnly = true)
    public ProjectResponse getProjectById(UUID workspaceId, UUID id) {
        Project project = getProject(workspaceId, id);
        return projectMapper.toResponse(project);
    }

    @Transactional
    public ProjectResponse updateProject(UUID workspaceId, UUID id, @Valid UpdateProjectRequest request) {
        workspaceService.getWorkspaceForMutation(workspaceId);
        Project project = getProject(workspaceId, id);
        requireMutable(project);

        project.setName(request.name());
        project.setDescription(request.description());
        return projectMapper.toResponse(project);
    }

    @Transactional
    public ProjectResponse updateProjectStatus(UUID workspaceId, UUID id, @Valid UpdateProjectStatusRequest request) {
        workspaceService.getWorkspaceForMutation(workspaceId);
        Project project = getProject(workspaceId, id);
        
        requireMutable(project);
        if (!project.canTransitionTo(request.status())) {
            throw new BusinessRuleException("Invalid project status transition: " + project.getStatus() + " -> " + request.status());
        }

        project.setStatus(request.status());
        return projectMapper.toResponse(project);
    }

    @Transactional
    public void deleteProject(UUID workspaceId, UUID id) {
        workspaceService.getWorkspaceForMutation(workspaceId);
        Project project = getProject(workspaceId, id);
        requireMutable(project);

        projectRepository.delete(project);
    }

    @Transactional(readOnly = true)
    public Project getProject(UUID workspaceId, UUID projectId) {
        workspaceService.getWorkspaceIfMember(workspaceId);
        return projectRepository.findByIdAndWorkspaceId(projectId, workspaceId)
                .orElseThrow(() -> new ProjectNotFoundException("Project not found"));
    }

    public void requireMutable(Project project) {
        if (project.isArchived()) {
            throw new BusinessRuleException("Archived projects cannot be modified");
        }
    }
}
